### PR TITLE
`Exam mode`: Change test case default visibility to after results release date

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Description
Finishing the work that was started in https://github.com/ls1intum/Artemis/pull/7738

* Adjusted the default test case visibility depending on whether the exercise is an exam or course exercise _(and corresponding translations)_
* Adjusted reset method to reset exam exercise test case visibilities to "After Release Date of Results"
* Adding JUnit test for new reset and creation of new exercises
* Fixing E2E tests that failed due to changes (Cypress and playwright)
* Fixing issues introduced by #7242 _(console error when loading student exam as instructor, change submitted state displayed twice)_
* Fixed issue in supporting_script introduced in #7379 _(`log.info` does not exist and lead to failing course setup script)_

### Steps for Testing
Prerequisites:
- 1 Instructor
- 1 Students

#### Create a programming exercise
1. `instructor`: create a course programming exercise
2. `instructor`: navigate to grading of the exercise and see that the visibility of the test cases is `Always`
3. `instructor`: create an exam programming exercise
4. `instructor`: navigate to grading of the exercise and see that the visibility of the test cases is `After Release Date of Results`
5. `instructor`: make some changes and see that `reset` rests to the expected default state (= test case visibility `After Release Date of Results`)
6. `student`: participate in the course exercise and see that you get immediate feedback
7. `student`: participate in the exam and see that you do only get the info whether your solution compiled

#### Import an exercise into an exam
1. `instructor`: import an exercise into an exam, see that the test case visibility is not changed by the import 
2. `instructor`: import the same exercise again into the exam (at least one test case visibility should be `always`) - this time toggle the option at the bottom in the `Assessment` section to set the visibility of test cases to `After Release Date of Results`, and see that the test case visibility is set to `After Release Date of Results`

For Importing exercise groups no modification/modification option applies, due to the assumption that exercise groups from other exams will be configured properly.

Not supported for imports from files, as there are further issues with this functionality right now that seem out of scope to me, see https://github.com/ls1intum/Artemis/issues/8562

#### Review the documentation
Have a look at the documentation from the build results
1. See that it is build properly https://artemis-platform--8451.org.readthedocs.build/en/8451/user/exams/instructors_guide.html#add-exercises
2. Is the newly added part understandable for you?

#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

#### Client
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [programming-exercise-task.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.ts.html) | 92.68% | ✅ |
| [programming-exercise-lifecycle.component.ts](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Client-Tests/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts.html) | 97.97% | ✅ |
| programming-exercises.ts | not found (added) | ❌  (just a type definition ) | 

#### Server
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|---------------------:|
| [ProgrammingExercise.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.domain/ProgrammingExercise.html) | 88% | ✅ ❌|
| [ProgrammingExerciseFeedbackCreationService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseFeedbackCreationService.html) | 94% | ✅ |
| [ProgrammingExerciseImportService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseImportService.html) | 93% | ✅ |
| [ProgrammingExerciseTestCaseService.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.service.programming/ProgrammingExerciseTestCaseService.html) | 89% | ✅ |
| [ProgrammingExerciseExportImportResource.java](https://bamboo.ase.in.tum.de/browse/ARTEMIS-TESTS6469/latest/artifact/shared/Coverage-Report-Server-Tests/de.tum.in.www1.artemis.web.rest.programming/ProgrammingExerciseExportImportResource.html) | 91% | ✅ |

### Screenshots
| Change                                                                              | Before                                                                                             | Now                                                                                                |
|-------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------|
| New default test visibility for new exam exercises                                  | image | image |
| Added import option (only when importing into exam, not available for file imports) | image | image |

## Summary by CodeRabbit

- **New Features**
	- Introduced default visibility settings for programming exercise test cases, varying between course and exam exercises.
	- Enhanced tooltip content in programming exercise grading components to reflect the exercise type.

- **Enhancements**
	- Improved logic for handling null values in student exam detail components.
	- Updated test case visibility management based on new default settings in programming exercises.

- **Bug Fixes**
	- Fixed issues with submission status toggling in student exam details.

- **Documentation**
	- Updated German and English localization files with new terms related to programming exercise visibility.

- **Tests**
	- Enhanced Cypress tests with new timeout settings and method adjustments for better reliability during assessments and exam setups.

### CODE REVIEW

**Comment by florian-glombik** [comment]
The failed client test seems to be flaky, runs through locally 

FAIL src/test/javascript/spec/component/exam/manage/student-exams/student-exam-timeline.component.spec.ts (8.402 s, 1362 MB heap size)
  ● Student Exam Timeline Component › should correctly set the values onPageChange

**Comment by florian-glombik** [comment]
The failed client test seems to be flaky, runs through locally 

```
FAIL src/test/javascript/spec/component/exam/manage/student-exams/student-exam-timeline.component.spec.ts (8.402 s, 1362 MB heap size)
   Student Exam Timeline Component › should correctly set the values onPageChange
```

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html
Comment (Strohgelaender) [inline src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:123]: Was this intentional?

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html
Comment (Strohgelaender) [inline src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html:108]: Maybe this could be more readable if you only have one transaation pipe after the selection at the end?

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html
Comment (florian-glombik) [reply src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:123]: Yes, I do regard 
```
  @if (!studentExam.submitted) {
      <span jhiTranslate="entity.action.toggleToSubmitted"></span>
  } @else {
      <span jhiTranslate="entity.action.toggleToUnsubmitted"></span>
  }
```

as a cleaner solution than the current implementation
```
  @if (!studentExam.submitted) {
      <span jhiTranslate="entity.action.toggleToSubmitted"></span>
  } @else {
      @if (studentExam.submitted) {
          <span jhiTranslate="entity.action.toggleToUnsubmitted"></span>
      }
  }
```

This seems to be a leftover from the angular migration, and no ngElse was used in this place. Do I overlook something?

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html:108]: My understanding of the use of pipes does not seem to be sufficient here, could you supply a code suggestion of how you would imagine it? 

For my understanding the pipe in line 99 is needed to pass parameters to the translation, and in 101 still the placeholder needs to be translated, so there is the need for a second usage of the pipe.
The same applies for 103-107. 

Is there an easier way to solve it?

Comment (Strohgelaender) [reply src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html:108]: Oh I did not notice the placeholder before, but thought that all of these were different cases. Is the placeholder really needed? Could you maybe integrate the string directly into the translation?

File: src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html
Comment (Strohgelaender) [reply src/main/webapp/app/exam/manage/student-exams/student-exam-detail.component.html:123]: I did not check the context of the code, just saw something that seem to not fit into the test case visibilty changes. This makes sense!

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-grading-tasks-table.component.html:108]: My intention behind the placeholders was to ensure the translation of a `Visibility-Option` is changed and that it is consistent in the description.

But the translation itself can be changed, of course, as it looks like
```
"visibility": "Feedback und Punkte für Testfälle, die mit „{{ afterDueDate }}“ markiert sind, werden erst danach für Studierende angezeigt. Nie sichtbare Tests werden nicht in die Punktebewertung einbezogen.",
"visibilityExam": "Feedback und Punkte für Testfälle, die mit „{{ afterReleaseDateOfResults }}“ markiert sind, werden erst danach nach der Veröffentlichung der Klausurergebnisse angezeigt. Nie sichtbare Tests werden nicht in die Punktebewertung einbezogen.",
```

Do you think that simplicity in the code is a better trade-off than the current solution?

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseTestCaseService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseTestCaseService.java:124]: Would it be better to only pass the actual exercise as parameter here?

File: src/main/webapp/i18n/de/programmingExercise.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/de/programmingExercise.json:314]:
```suggestion
                    "visibilityExam": "Feedback und Punkte für Testfälle, die mit „{{ afterReleaseDateOfResults }}“ markiert sind, werden erst nach der Veröffentlichung der Klausurergebnisse angezeigt. Nie sichtbare Tests werden nicht in die Punktebewertung einbezogen.",
```

File: src/main/webapp/i18n/en/programmingExercise.json
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/programmingExercise.json:316]: Why is there a translation for afterReleaseDateOfResults but not for ‘Never’?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/i18n/en/programmingExercise.json
Comment (florian-glombik) [reply src/main/webapp/i18n/en/programmingExercise.json:316]: No particular reason, because I did only change the part directly addressed by this PR for now.

Do you have an opinion on which approach is better, as I also had a discussion with @Strohgelaender  about this https://github.com/ls1intum/Artemis/pull/8451#discussion_r1573838850? Translation (less error prone) vs. hardcoded (less complex)?

Comment (florian-glombik) [reply src/main/webapp/i18n/en/programmingExercise.json]: It is now simplified https://github.com/ls1intum/Artemis/pull/8451/commits/fbe5708e89225f088c1e3e7f2bef0a8cd03cdcdc

File: docs/user/exams/instructor/add_exercises.png
Comment (florian-glombik) [inline docs/user/exams/instructor/add_exercises.png:1]: The button for creating a file upload exercise was missing in the previous screenshot

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html:187]: We want to use `artemisTranslate` here!

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:322]: Is that TODO for now or the future? Why might it need to be adjusted?

File: src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseImportService.java:309]:
```suggestion
            newProgrammingExercise.setTestCases(new HashSet<>(updatedTestCases));
```
Set.copyOf is unmodifiable, which could lead to problems in the future if we want to change test cases after this method

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-creation-config.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-creation-config.ts:51]:
```suggestion
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:1096]:
```suggestion
```

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:641]: TODO?

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html
Comment (JohannesStoehr) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html:136]: Why did you change that?

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

File: src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/manage/update/programming-exercise-update.component.ts:641]: This is outdated, the import from files does not work properly currently, before adding the support for that the import should be fixed (I had a short look at the issue and it seems out of scope to me, considering that the PR is already quite big) [Programming exercises: Build results not available after file import #8562](https://github.com/ls1intum/Artemis/issues/8562)

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html
Comment (florian-glombik) [reply src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html:136]: I misread the guidelines and thought we would choose `artemisTranslate` over `jhiTranslate`, which is actually the other way around

https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/client.html

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java
Comment (florian-glombik) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:322]: It is out of date for now and unintentionally left there - the last boolean value will need to be modified to support the new import option (currently it is always set to false, so the test case visibilities will not be overridden)

Due to further issues with the file import this is not implemented now. https://github.com/ls1intum/Artemis/issues/8562

File: src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java
Comment (florian-glombik) [reply src/main/java/de/tum/in/www1/artemis/service/exam/ExamImportService.java:323]: Fixing the file import, which would be needed to properly test the new import option is out of scope of this PR, therefore no reasonable dynamic value can be added here.

**Review by JohannesStoehr — APPROVED** [review]

**Review by BBesrour — APPROVED** [review]
Demonstrated in a testing session.
Tested again on TS3 and works as expected: importing an exam exercise doesn't change visibility unless `set the visibility` is enabled
Exam that was created for testing: https://artemis-test3.artemis.cit.tum.de/course-management/48/exams/7/exercise-groups

Documentation is clear and understanble

**Review by az108 — APPROVED** [review]
tested on ts3 with the link : https://artemis-test3.artemis.cit.tum.de/course-management/38/exams/9/exercise-groups

Works as expected and code LGTM

**Review by BBesrour — APPROVED** [review]
re-approve after merge conflicts

**Review by JohannesStoehr — APPROVED** [review]
Reapprove after merge

**Review by sarpsahinalp — APPROVED** [review]
Tested on ts3 works as expected!

**Review by az108 — APPROVED** [review]
reapprove

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.ts
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.ts:86]: We should check for type, too. -->  if (this.isExamExercise && value **===** Visibility.AfterDueDate) {

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:26]:
```suggestion
    @Input() importOptions?: ImportOptions;
```

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.html:149]: personal preference: I would create a toggle-function. Makes the html file bit cleaner. :)

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.ts
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.ts:86]: We should check for type, too. --> if (this.isExamExercise && value === Visibility.AfterDueDate) {

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:108]:
```suggestion
            this.urlSubscription?.unsubscribe();
```

File: src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.html
Comment (rabeatwork) [inline src/main/webapp/app/exercises/programming/manage/grading/tasks/programming-exercise-task/programming-exercise-task.component.html:55]: 🙏🏼

**Review by BBesrour — DISMISSED** [review]
Re-approve

**Review by az108 — APPROVED** [review]
reapprove

**Review by JohannesStoehr — DISMISSED** [review]
Code

**Review by rabeatwork — DISMISSED** [review]
Maintainer approved

File: src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts
Comment (florian-glombik) [inline src/main/webapp/app/exercises/programming/shared/lifecycle/programming-exercise-lifecycle.component.ts:135]: `toggleManualFeedbackRequests` was renamed to `allowFeedbackRequests` on `develop` - there was a conflict because the method above `toggleSetTestCaseVisibilityToAfterDueDate` was added in this PR (= change in the same line)

**Review by JohannesStoehr — APPROVED** [review]
Approve after merge conflict

**Review by BBesrour — APPROVED** [review]
Approve after merge conflict

**Pull request merged**
